wallet: Keep track of the wallet's own transaction outputs in memory

achow101 opened the pull request:
Currently, the wallet is not actually aware about its own transaction outputs. Instead, it will iterate all of the transactions stored in `mapWallet`, and then all of the outputs of those transactions, in order to figure out what belongs to it for the purposes of coin selection and balance calculation. For balance calculation, there is caching that results in it only iterating all of the transactions, but not all of the outputs. However when the cache is dirty, everything is iterated. This is especially problematic for wallets that have a lot of transactions, or transactions that have a lot of unrelated outputs (as may occur with coinjoins or batched payments).

This PR helps to resolve this issue by making the wallet track all of the outputs that belong to it in a new member `m_txos`. Note that this includes outputs that may have already been spent. Both balance calculation (`GetBalance`) and coin selection (`AvailableCoins`) are updated to iterate `m_txos`. This is generally faster since it ignores all of the unrelated outputs, and it is not slower as in the worst case of wallets containing only single output transactions, it's exactly the same number of outputs iterated.

`m_txos` is memory only, and it is populated during wallet loading. When each transaction is loaded, all of its outputs are checked to see if it is `IsMine`, and if so, an entry added to `m_txos`. When new transactions are received, the same procedure is done.

Since imports can change the `IsMine` status of a transaction (although they can only be "promoted" from watchonly to spendable), all of the import RPCs will be a bit slower as they re-iterate all transactions and all outputs to update `m_txos`.

Each output in `m_txos` is stored in a new `WalletTXO` class. This class contains references to the parent `CWalletTx` and the `CTxOut` itself. It also caches the `IsMine` value of the txout. This should be safe as `IsMine` should not change unless there are imports. This allows us to have additional performance improvements in places that use these `WalletTXO`s as they can use the cached `IsMine` rather than repeatedly calling `IsMine` which can be expensive.

The existing `WalletBalance` benchmark demonstrates the performance improvement that this PR makes. The existing `WalletAvailableCoins` benchmark doesn't as all of the outputs used in that benchmark belong to the test wallet. I've updated that benchmark to have a bunch of unrelated outputs in each transaction so that the difference is demonstrated.

This is part of a larger project to have the wallet actually track and store a set of its UTXOs.

Built on #24914 as it requires loading the txs last in order for `m_txos` to be built correctly.

***

## Benchmarks:

Master:

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       34,590,013.00 |               28.91 |    0.0% |  812,669,269.00 |  148,360,642.50 |  5.478 |  18,356,853.00 |    0.2% |      0.76 | `WalletAvailableCoins`
|            3,193.46 |          313,139.91 |    0.4% |       96,868.06 |       13,731.82 |  7.054 |      26,238.01 |    0.1% |      0.01 | `WalletBalanceClean`
|           26,871.18 |           37,214.59 |    3.3% |      768,179.50 |      115,544.39 |  6.648 |     154,171.09 |    0.1% |      0.01 | `WalletBalanceDirty`
|            3,177.30 |          314,732.47 |    0.2% |       96,868.06 |       13,646.20 |  7.099 |      26,238.01 |    0.1% |      0.01 | `WalletBalanceMine`
|               10.73 |       93,186,952.53 |    0.1% |          157.00 |           46.14 |  3.403 |          36.00 |    0.0% |      0.01 | `WalletBalanceWatch`
|      590,497,920.00 |                1.69 |    0.1% |12,761,692,005.00 |2,536,899,595.00 |  5.030 | 129,124,399.00 |    0.7% |      6.50 | `WalletCreateEncrypted`
|      182,929,529.00 |                5.47 |    0.0% |4,199,271,397.00 |  785,477,302.00 |  5.346 |  75,363,377.00 |    1.1% |      2.01 | `WalletCreatePlain`
|          699,337.20 |            1,429.93 |    0.7% |   18,054,294.00 |    3,005,072.20 |  6.008 |     387,756.60 |    0.3% |      0.04 | `WalletCreateTxUseOnlyPresetInputs`
|       32,068,583.80 |               31.18 |    0.5% |  562,026,110.00 |  137,457,635.60 |  4.089 |  90,667,459.40 |    0.3% |      1.78 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|               36.62 |       27,306,578.40 |    0.5% |          951.00 |          157.05 |  6.056 |         133.00 |    0.0% |      0.01 | `WalletIsMineDescriptors`
|               35.00 |       28,569,989.42 |    0.7% |          937.00 |          150.33 |  6.233 |         129.00 |    0.0% |      0.01 | `WalletIsMineMigratedDescriptors`
|      203,284,889.00 |                4.92 |    0.0% |4,622,691,895.00 |  872,875,275.00 |  5.296 |  90,345,002.00 |    1.2% |      1.02 | `WalletLoadingDescriptors`
|    1,165,766,084.00 |                0.86 |    0.0% |24,139,316,211.00 |5,005,218,705.00 |  4.823 |2,664,455,775.00 |    0.1% |      1.17 | `WalletMigration`

PR:

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       33,975,750.50 |               29.43 |    0.1% |  794,719,150.50 |  145,763,550.00 |  5.452 |  16,036,630.50 |    0.2% |      0.75 | `WalletAvailableCoins`
|            2,442.01 |          409,498.46 |    0.2% |       60,782.04 |       10,500.60 |  5.788 |       9,492.01 |    0.3% |      0.01 | `WalletBalanceClean`
|            2,763.12 |          361,909.21 |    0.2% |       61,493.05 |       11,859.48 |  5.185 |       9,625.01 |    0.2% |      0.01 | `WalletBalanceDirty`
|            2,347.98 |          425,898.72 |    0.3% |       60,782.04 |       10,082.73 |  6.028 |       9,492.01 |    0.2% |      0.01 | `WalletBalanceMine`
|               11.67 |       85,654,630.36 |    0.2% |          176.00 |           50.18 |  3.508 |          40.00 |    0.0% |      0.01 | `WalletBalanceWatch`
|      590,119,519.00 |                1.69 |    0.1% |12,754,398,258.00 |2,534,998,522.00 |  5.031 | 129,078,027.00 |    0.7% |      6.50 | `WalletCreateEncrypted`
|      183,124,790.00 |                5.46 |    0.1% |4,199,212,926.00 |  786,323,886.00 |  5.340 |  75,354,437.00 |    1.1% |      2.02 | `WalletCreatePlain`
|          669,643.00 |            1,493.33 |    0.1% |   17,213,904.20 |    2,877,336.40 |  5.983 |     394,292.80 |    0.3% |      0.04 | `WalletCreateTxUseOnlyPresetInputs`
|       26,205,987.80 |               38.16 |    0.8% |  365,551,340.80 |  112,376,905.20 |  3.253 |  65,684,276.20 |    0.4% |      1.44 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|               34.75 |       28,778,846.38 |    0.1% |          937.00 |          149.41 |  6.271 |         129.00 |    0.0% |      0.01 | `WalletIsMineDescriptors`
|               29.91 |       33,428,072.85 |    0.2% |          920.00 |          128.63 |  7.152 |         126.00 |    0.0% |      0.01 | `WalletIsMineMigratedDescriptors`
|      202,437,985.00 |                4.94 |    0.1% |4,626,686,256.00 |  869,439,274.00 |  5.321 |  90,961,305.00 |    1.1% |      1.02 | `WalletLoadingDescriptors`
|    1,158,394,152.00 |                0.86 |    0.0% |24,143,589,972.00 |4,971,946,380.00 |  4.856 |2,665,355,654.00 |    0.1% |      1.16 | `WalletMigration`

Xekyo commented:
Concept ACK

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -217,133 +217,141 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+            tx_safe_cache[outpoint.hash] = {true, safeTx};
+        }
+        const auto& [tx_ok, tx_safe] = tx_safe_cache.at(outpoint.hash);
+        Assert(tx_ok);
```
In https://github.com/bitcoin/bitcoin/commit/881650ed319c6ed74e66ae56c34cdee93125231b " wallet: Use wallet's TXO set in AvailableCoins "

Can't `tx_ok` still be false at this point if a transaction that has a depth of 0 and is not in the mempool has two txos that belong to the wallet? Since then when the first txo is evaluated, a `{false, false}` entry will be created for that transaction hash, and then when the second txo is evaluated, that entry will be used because the mempool check will be skipped?

achow101 replied:
Good point, changed to an `if`.

achow101 replied:
Indeed. I've left the `if` there anyways as a belt-and-suspenders, and changed it to also `Assume`.

ishaanam reviewed (commented):
Concept ACK

Still working on reviewing this, but here are some initial things. Thanks for breaking it up into twelve commits, it makes it easy to review.

ishaanam commented:
The commit message for 68bb1463dc1f995fbfdb75d3acef625bde104275 should be updated because now that #27145 has gotten merged, the described issue has already been fixed.

remyers commented:
Concept ACK

My preliminary testing confirmed that this PR will significantly speed up common operations like `fundrawtransaction` on wallets with many years of transaction history. A single `fundrawtransaction` RPC call that took 5 sec on a simulated large wallet now takes 50ms. Services providers with large wallets that need to make many RPC calls per minute will benefit from reduced lock contention on `cs_wallet`; for example, LSPs using bitcoin core wallets like Eclair.

Caching the wallet's tx outputs in memory seems like a straight forward solution as long as the cache of tx outs in `m_txos` are always updated (or regenerated) if the underlying set of wallet txs in `mapWallet` or the `m_spk_managers` change.

remyers commented:
Exposing `RefreshWalletTXOs()` as a public method puts the responsibility of knowing when it should be called outside of the `CWallet` scope. What do you think about making it private and clearing `m_txos` (or adding a new flag) to trigger `RefreshWalletTXOs()` at query time?

This alternate approach would require `GetTXO(const COutPoint& outpoint)` and `GetTXOs()` to first call `RefreshWalletTXOs()` if `m_txos` is empty.  There are only a few methods that change `m_spk_managers` that would need to clear `m_txos` to trigger a one time `RefreshWalletTXOs()`:
 - `ImportPrivKeys`
 - `ImportPublicKeys`
 - `ImportScriptPubKeys`
 - `AddScriptPubKeyMan`

It's also only a small set of places where keys are imported in `backup.cpp`, so the question of leaving `RefreshWalletTXOs()` public might just be a matter of which approach would be less foot-gunny in the future and if there are disadvantages to delaying the regeneration of txos.

I think you can safely make `RefreshWalletTxTXOs(const CWalletTx& wtx)` a private method now though because it is only called by wallet methods that add a wtx to `mapWallet` like `LoadToWallet` and `AddToWallet`.

achow101 commented:
> Exposing `RefreshWalletTXOs()` as a public method puts the responsibility of knowing when it should be called outside of the `CWallet` scope. What do you think about making it private and clearing `m_txos` (or adding a new flag) to trigger `RefreshWalletTXOs()` at query time?

I'm not sure that's meaningfully different. The caller still needs to know when to clear `m_txos` in order to trigger a future `RefreshWalletTXOs()`. The way it is now also provides a better UX as the slow part of refreshing the TXOs occurs during slow import operations rather than the first time those TXOs are needed after the import.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+        for (const auto& [outpoint, txo] : wallet.GetTXOs()) {
+            Assert(MoneyRange(txo.GetTxOut().nValue));
+
+            const bool is_trusted{CachedTxIsTrusted(wallet, txo.GetWalletTx(), trusted_parents)};
```
Done

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                } else if (is_trusted && tx_depth >= min_depth) {
+                    ret.m_mine_trusted += credit_mine;
+                    ret.m_watchonly_trusted += credit_watchonly;
+                } if (!is_trusted && tx_depth == 0 && txo.GetWalletTx().InMempool()) {
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -480,6 +480,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     const std::unordered_map<COutPoint, WalletTXO, SaltedOutpointHasher>& GetTXOs() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) { AssertLockHeld(cs_wallet); return m_txos; };
 
     void RefreshWalletTxTXOs(const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    void RefreshWalletTXOs() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Hmm, good point. The expected usage of these is that only new keys will be generated/added, so it shouldn't have any effect on the existing transactions in the wallet. However, I suppose they could be used in a way that does result in existing txs having their IsMine changed. I've also added this to `addmultisigaddress` and `newkeypool` since they could be used in the same way.

achow101 commented on src/wallet/rpc/backup.cpp:
```diff
@@ -627,6 +631,7 @@ RPCHelpMan importwallet()
 
             progress++;
         }
+        pwallet->RefreshWalletTXOs();
```
The wallet may not rescan every transaction in the wallet since it does look at the given key birthdays.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -217,133 +217,141 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-            const COutPoint outpoint(wtxid, i);
+        if (mine == ISMINE_NO) {
+            continue;
+        }
```
Done

remyers commented:
> I'm not sure that's meaningfully different. The caller still needs to know when to clear `m_txos` in order to trigger a future `RefreshWalletTXOs()`. The way it is now also provides a better UX as the slow part of refreshing the TXOs occurs during slow import operations rather than the first time those TXOs are needed after the import.

Just to clarify, I mean that a reviewer, but not method caller, would need to make sure every method in CWallet that invalidates `m_txos` also clears it. I agree with your point though that forcing the refresh at import time makes sense. I also notice that there are a number of places where a `ScriptPubKeyMan` gets added or updated outside of CWallet, so I understand now why it's not really possible to make `RefreshWalletTXOs()` something internal to `CWallet`.

ishaanam reviewed (commented):
ACK 9199832d4440a875a150aa3ba3bc1bc201d3fd72

I've reviewed everything but the tests and benchmarks.

achow101 commented on test/functional/wallet_abandonconflict.py:
```diff
@@ -231,7 +231,8 @@ def run_test(self):
 
-        # Invalidate the block with the double spend. B & C's 10 BTC outputs should no longer be available
+        # Invalidate the block with the double spend and both B and C's 10 BTC output should no longer be available
+        # AB1 and ABC2's outputs are also not available since neither of those txs are in the mempool
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -504,6 +504,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     const std::unordered_map<COutPoint, WalletTXO, SaltedOutpointHasher>& GetTXOs() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) { AssertLockHeld(cs_wallet); return m_txos; };
 
     void RefreshWalletTxTXOs(const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    void RefreshWalletTXOs() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Renamed as suggested.

achow101 commented on src/bench/wallet_create_tx.cpp:
```diff
@@ -47,10 +47,17 @@ void generateFakeBlock(const CChainParams& params,
+    // Fill the coinbase with outputs that don't belong to the wallet in order to benchmark
+    // AvailableCoins' behavior with unnecessary TXOs
+    for (int i = 0; i < 50; ++i) {
+        coinbase_tx.vout.push_back(CTxOut(1 * COIN / 50, CScript(OP_TRUE)));
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1555,16 +1555,10 @@ void CWallet::BlockUntilSyncedToCurrentChain() const {
 // and a not-"is mine" (according to the filter) input.
 CAmount CWallet::GetDebit(const CTxIn &txin, const isminefilter& filter) const
```
"Debit" refers to the amount leaving the wallet, assuming the UTXO spent by the input matches the filter. "Debit" operates on a transaction's inputs, while "credit" operates on the outputs.

yancyribbens replied:
Is this documented anywhere?  This is a good question and I would imagine others will have the same in the future.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -256,6 +256,10 @@ bool CachedTxIsFromMe(const CWallet& wallet, const CWalletTx& wtx, const isminef
     AssertLockHeld(wallet.cs_wallet);
+
+    // This wtx is already trusted
+    if (trusted_parents.count(wtx.GetHash()) > 0) return true;
```
Yes, and that does happen, e.g. send to self transactions. The tests also do this a lot.

murchandamus commented on src/wallet/spend.cpp:
```diff
@@ -320,137 +320,145 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-        if (nDepth == 0 && wtx.mapValue.count("replaces_txid")) {
-            safeTx = false;
-        }
+        if (wallet.IsSpent(outpoint))
+            continue;
```
I have the impression that `m_txos` is also used to look up information on old transactions and we therefore want to keep all transaction outputs the wallet ever had.

Would it perhaps make sense to also have a reduced set that only keeps the potentially _unspent_ TXOs? I could see a wallet that participates in a few dozen transactions every day have magnitudes more TXOs than UTXOs. So perhaps the work the wallet has to perform every time it builds a transaction could be shortcut by storing e.g. `m_is_mature_spent` on `WalletTXO` which could be set to true when the transaction spending an output has 100 confirmations, and then either separately keeping a list of all TXOs that are not `m_is_mature_spent` or at least iterate over them more quickly directly on the basis of that member, without needing to have complex reorg handling to keep this collection up to date.

murchandamus replied:
Ah, nevermind, I just saw https://github.com/bitcoin/bitcoin/pull/27865

murchandamus reviewed (commented):
ACK 0c0022311a0cffc4c03b3763c9b60f93ede13687

While I am pretty sure that the reorg logic should handle any transactions getting added or removed from the wallet, it might be useful to add a test that the `m_txos` set would be correct after a reorg, if that’s not obvious.

aureleoules commented on src/wallet/receive.cpp:
```diff
@@ -293,27 +252,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                } else if (is_trusted && tx_depth >= min_depth) {
+                    ret.m_mine_trusted += credit_mine;
+                    ret.m_watchonly_trusted += credit_watchonly;
+                } if (!is_trusted && txo.GetWalletTx().InMempool()) {
```
Seems like a hazard
```suggestion
                } else if (!is_trusted && txo.GetWalletTx().InMempool()) {
```

achow101 replied:
Good catch, fixed.

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -320,137 +316,145 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-            // Skip manually selected coins (the caller can fetch them directly)
-            if (coinControl && coinControl->HasSelected() && coinControl->IsSelected(outpoint))
-                continue;
+        if (tx_safe_cache.count(outpoint.hash) == 0 ) {
```
nit: Could use `contains` (and other places too)
```suggestion
        if (!tx_safe_cache.contains(outpoint.hash)) {
```

achow101 replied:
Done

murchandamus reviewed (commented):
ReACK https://github.com/bitcoin/bitcoin/commit/39f0ab3e126704b30f7ed2b86690f60b631d2ec4 compared to my prior review, there are only the changes mentioned in response to @aureleoules.

achow101 committed:
walletdb: Load Txs last

Need to load txs last so that IsMine works.

achow101 committed:
wallet: Keep track of transaction outputs owned by the wallet

When loading transactions to the wallet, check the outputs, and keep
track of the ones that are IsMine.

achow101 committed:
wallet: Exit IsTrustedTx early if wtx is already in trusted_parents

achow101 committed:
wallet: Change balance calculation to use m_txos

Since we track the outputs owned by the wallet with m_txos, we can now
calculate the balance of the wallet by iterating m_txos and summing up
the amounts of the unspent txos.

As ISMINE_USED is not an actual isminetype that we attach to outputs and
was just passed into `CachedTxGetAvailableCredit` for convenience, we
pull the same determining logic from that function into `GetBalances` in
order to preserve existing behavior.

achow101 committed:
wallet: Recalculate the wallet's txos after any imports

achow101 committed:
test: Test for balance update due to untracked output becoming spendable

achow101 committed:
wallet: Use wallet's TXO set in AvailableCoins

Instead of iterating every transaction and every output stored in wallet
when trying to figure out what outputs can be spent, iterate the TXO set
which should be a lot smaller.

achow101 committed:
wallet: Retrieve TXO directly in FetchSelectedInputs

Instead of searching mapWallet for the preselected inputs, search
m_txos.

wallet_fundrawtransaction.py spends external inputs and needs the change
output to also belong to the test wallet for the oversized tx test.

achow101 committed:
wallet: Recompute wallet TXOs after descriptor migration

When a legacy wallet has been migrated to contain descriptors, but
before the transactions have been updated to match, we need to recompute
the wallet TXOs so that the transaction update will work correctly.

achow101 committed:
wallet: Have GetDebit use the wallet's TXO set

Instead of looking up the previous tx in mapWallet, and then calling
IsMine on the specified output, use the TXO set and its cached IsMine
value.

achow101 committed:
wallet: Remove unused CachedTxGet{Available,Immature}Credit

These two functions are no longer used as GetBalances now uses the TXO
set rather than per-tx cached balances

achow101 committed:
bench: Have AvailableCoins benchmark include a lot of unrelated utxos

One of the main issues with AvailableCoins is its performance when txs
have unrelated outputs, so update the benchmark to check the performance
of that.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4467,4 +4475,22 @@ void CWallet::TopUpCallback(const std::set<CScript>& spks, ScriptPubKeyMan* spkm
+            m_txos.at(outpoint).SetIsMine(ismine);
+        } else {
+            m_txos.emplace(outpoint, WalletTXO{wtx, txout, ismine});
+        }
```
Outputs don't always stay spent. Replacements, reorgs, and eviction could all make a transaction no longer be "IsSpent", so checking that here would require a lot of other structural changes to ensure that it is correct.

Storing the value of `IsSpent` could be an alternative to #27865, however I would still leave that for a followup. Any change that tries to actually have a proper UTXO set for the wallet can have unexpected side effects and/or things we didn't notice so I want to keep that separate for now. This PR is to setup for that by making the wallet operate on a TXO set.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+            const bool is_trusted{CachedTxIsTrusted(wallet, txo.GetWalletTx())};
+            const int tx_depth{wallet.GetTxDepthInMainChain(txo.GetWalletTx())};
+
+            if (!wallet.IsSpent(outpoint) && (allow_used_addresses || !wallet.IsSpentKey(txo.GetTxOut().scriptPubKey))) {
```
`ISMINE_USED` is not actually an isminetype that is returned by `IsMine`, so it will not be set in the `txo.GetIsMine()`'s result. It's really just a hack for it to be an isminetype so it can be passed as part of various ismine filters, but these always check that separately from the actual `IsMine` result.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                } else if (txo.GetIsMine() == ISMINE_WATCH_ONLY) {
+                    credit_watchonly = txo.GetTxOut().nValue;
+                } else {
+                    // We shouldn't see any other isminetypes
+                    Assume(false);
```
No. `ISMINE_USED` is never actually returned from `IsMine`.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                    Assume(false);
+                }
+
+                // Set the amounts in the return object
```
Can you be more specific as to what kind of transaction you think would result in a different balance? Nothing in this PR should change behavior.

achow101 commented on test/functional/wallet_balance.py:
```diff
@@ -340,5 +375,30 @@ def test_balances(*, fee_node_1=0):
+        balances = wallet.getbalances()
+        assert_equal(balances["mine"]["trusted"], amount)
+
+        # Don't rescan to make sure that the import updates the wallet txos
+        wallet.importprivkey(privkey=import_key2.privkey, rescan=False)
+        balances = wallet.getbalances()
+        assert_equal(balances["mine"]["trusted"], amount * 2)
```
That is correct.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -320,137 +320,145 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-    {
-        const uint256& txid = entry.first;
-        const CWalletTx& wtx = entry.second;
+    // Cache for whether each tx passes the tx checks (first bool), and whether the transaction is "safe" (second bool)
```
The checks are already explained in detail below where they are actually checked. "tx checks" just refers to the checks that occur per tx rather than per output. I don't think it is useful to duplicate the explanations here as well.

I've added a comment that should make it easier to find where the tx checks are performed so you can read the comments that already exist.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1143,6 +1143,9 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
     // Break debit/credit balance caches:
     wtx.MarkDirty();
 
+    // Get the outputs that belong to the wallet
```
Changed to "cache".

"on the wallet's transactions" does not make the comment valid English sentence, so I'm not sure what meaning you were trying to get across.

murchandamus replied:
I think I wanted to express that this function operates only on the transactions that the wallet tracks already.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4467,4 +4475,22 @@ void CWallet::TopUpCallback(const std::set<CScript>& spks, ScriptPubKeyMan* spkm
+            continue;
+        }
+        COutPoint outpoint(wtx.GetHash(), i);
+        if (m_txos.count(outpoint) > 0) {
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -425,6 +425,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     //! Cache of descriptor ScriptPubKeys used for IsMine. Maps ScriptPubKey to set of spkms
     std::unordered_map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks;
 
+    //! Set of transaction outputs owned by this wallet
```
Done

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -256,6 +256,10 @@ bool CachedTxIsFromMe(const CWallet& wallet, const CWalletTx& wtx, const isminef
     AssertLockHeld(wallet.cs_wallet);
+
+    // This wtx is already trusted
+    if (trusted_parents.count(wtx.GetHash()) > 0) return true;
```
Done

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
 {
     Balance ret;
-    isminefilter reuse_filter = avoid_reuse ? ISMINE_NO : ISMINE_USED;
+    bool allow_used_addresses = !avoid_reuse || !wallet.IsWalletFlagSet(WALLET_FLAG_AVOID_REUSE);
```
Mentioned in the commit message.

achow101 commented:
Rebased for silent merge conflict.

murchandamus reviewed (commented):
ReACK 3b6ac646ffea930ae7d8c5d93acaba3587cc9dcd

achow101 commented on src/wallet/wallet.h:
```diff
@@ -507,6 +510,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     std::set<uint256> GetTxConflicts(const CWalletTx& wtx) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    void RefreshWalletTxTXOs(const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -513,6 +513,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     const std::unordered_map<COutPoint, WalletTXO, SaltedOutpointHasher>& GetTXOs() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) { AssertLockHeld(cs_wallet); return m_txos; };
 
     void RefreshWalletTxTXOs(const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    void RefreshAllTXOs() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Done

jonatack commented:
Concept ACK, if this patchset can meaningfully improve performance and reliably refresh wallet utxos to keep the in-memory cache in sync.

murchandamus commented:
ReACK 48545bdb3b9f59a8b78640dbf20afff6e2663ad7

murchandamus reviewed (commented):
ReACK 48545bdb3b9f59a8b78640dbf20afff6e2663ad7

mahmoudeA reviewed (commented):
src/wallet/wallet.h

mahmoudeA commented:
src/wallet/wallet.h

jonatack commented:
I plan to review this soon.

yancyribbens commented on src/wallet/wallet.cpp:
```diff
@@ -4621,4 +4627,40 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
     return std::nullopt;
 }
+
+void CWallet::RefreshSingleTxTXOs(const CWalletTx& wtx)
```
This method name looks funky. SingleTxTXO is short for single transaction transaction output?

achow101 replied:
Yes. It is a single transaction's transaction outputs.

achow101 replied:
Renamed to `RefreshTXOsFromTx`.

w0xlt commented:
This PR also simplifies the `GetBalance` and `AvailableCoins` methods. 
The introduction of the `GetTXO()` method also makes the wallet interface clearer. 
As far as I could verify, the `m_txo` in-memory  cache is reliably kept in sync.

ACK https://github.com/bitcoin/bitcoin/pull/27286/commits/48545bdb3b9f59a8b78640dbf20afff6e2663ad7

rkrux commented:
> I plan to review this soon.

Same, I intend to review it this month.

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/27286/commits/26014804bb792bde141baf3e9d9ac30b6a27c988

achow101 commented on src/wallet/wallet.h:
```diff
@@ -514,6 +514,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     /** Cache outputs that belong to the wallet from a single transaction */
     void RefreshSingleTxTXOs(const CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    /** Cache outputs that belong to the wallt for all tranasctions in the wallet */
```
Fixed

w0xlt commented on src/wallet/receive.cpp:
```diff
@@ -257,6 +257,10 @@ bool CachedTxIsFromMe(const CWallet& wallet, const CWalletTx& wtx, const isminef
     AssertLockHeld(wallet.cs_wallet);
+
+    // This wtx is already trusted
+    if (trusted_parents.contains(wtx.GetHash())) return true;
```
Is this optimization directly related to this PR, or did you take the opportunity to include it here?

achow101 replied:
I believe it's necessary so that repeated calls to `CachedTxIsTrusted` that would now occur once for each TXO can be skipped early.

w0xlt commented on test/functional/wallet_balance.py:
```diff
@@ -285,5 +287,30 @@ def test_balances(*, fee_node_1=0):
 
+        self.log.info("Test that the balance is updated by an import that makes an untracked output in an existing tx \"mine\"")
+        default = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+        self.nodes[0].createwallet("importupdate")
```
Is this wallet watch-only (per commit description) ?

achow101 replied:
It is not. That part of the test no longer exists after #28710. I've updated the commit message.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/27286/commits/51d4846f648e79e3e098f1f1bea7d5e823ebdea8

murchandamus commented on src/wallet/rpc/backup.cpp:
```diff
@@ -401,6 +401,7 @@ RPCHelpMan importdescriptors()
             }
         }
         pwallet->ConnectScriptPubKeyManNotifiers();
+        pwallet->RefreshAllTXOs();
```
In "wallet: Recalculate the wallet's txos after any imports" (850de8343e36a17a5c875b1be974153bf06efaca):
I’m surprised that the prior commit did not have failing tests if the addition of `RefreshAllTXOs()` is necessary here. Is there maybe test coverage missing here?

achow101 replied:
The test coverage is added in the next commit "test: Test for balance update due to untracked output becoming spendable" which I have now moved to be earlier in the PR.

murchandamus commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,40 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
-            if (!is_trusted && tx_depth == 0 && wtx.InMempool()) {
-                ret.m_mine_untrusted_pending += tx_credit_mine;
-                ret.m_watchonly_untrusted_pending += tx_credit_watchonly;
+        for (const auto& [outpoint, txo] : wallet.GetTXOs()) {
```
I think it would make more sense to me if "wallet: Change balance calculation to use m_txos" (ff73562fb5df406da95d6a6327e3d22b8d3e12f2) came after "test: Test for balance update due to untracked output becoming spendable" (f062a438620625bfa0d20022e76cad1705587737).

First the fix, then the test, then the refactor passes the test.

achow101 replied:
Done.

This actually revealed a bug in `importdescriptors`, I've added a commit to fix that, although it should actually be moot after the changes in this PR.

murchandamus reviewed (commented):
ACK 51d4846f648e79e3e098f1f1bea7d5e823ebdea8

achow101 committed:
wallet: MarkDirty after AddWalletDescriptor

After adding a wallet descriptor (typically by import), mark all balance
caches dirty. This allows transactions that the wallet already knows
about that have outputs that are now ISMINE_SPENDABLE after the import
to actually be shown in balance calculations. Legacy wallet imports
would do this, but importdescriptors did not.

achow101 commented on src/wallet/transaction.h:
```diff
@@ -369,6 +369,28 @@ struct WalletTxOrderComparator {
+    : m_wtx(wtx),
+    m_output(output),
+    m_ismine(ismine)
+    {}
```
Done as suggested

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4454,4 +4462,22 @@ std::optional<CKey> CWallet::GetKey(const CKeyID& keyid) const
+        if (ismine == ISMINE_NO) {
+            continue;
+        }
+        COutPoint outpoint(wtx.GetHash(), i);
+        if (m_txos.contains(outpoint)) {
+            m_txos.at(outpoint).SetIsMine(ismine);
+        } else {
+            m_txos.emplace(outpoint, WalletTXO{wtx, txout, ismine});
+        }
```
As-is is more legible, and this area of code changes significantly in #27865 where having it like it is now makes those changes easier to read.

davidgumberg replied:
Agreed

murchandamus commented:
reACK https://github.com/bitcoin/bitcoin/pull/27286/commits/29cf059e6e592f7f82b982493a043219cdfa5b40
    
via git range-diff efac285a0d709f48856c37b15bfa09af94c1d75b..51d4846f648e79e3e098f1f1bea7d5e823ebdea8 b15c386933eb7bc5ed8644dd37f8169d81ea5a22..29cf059e6e592f7f82b982493a043219cdfa5b40

Hooray for `Txid` class!

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/27286/commits/29cf059e6e592f7f82b982493a043219cdfa5b40

rkrux reviewed (commented):
I have just started reviewing this PR - have gone through the PR description & commit messages, going through the diff at the moment.

The intent of the PR is quite cool! As I understand, the motivation comes from this issue #27002. Due to the nature of the change, I decided to run benchmarks on my system and have shared below the best results of the lot I had (the results seem to be consistent across multiple runs).

### Observations
1. I can clearly see noticeable performance improvements in the `WalletBalance` benchmarks.
2. However, both `WalletAvailableCoins` & `WalletCreateTxUsePresetInputsAndCoinSelection` show a marginal decline instead with decent fluctuation (err %). 
3. Interestingly, the `WalletIsMine*` benchmarks showed improvement when I ran the bench with default value of 10ms, but declined when I ran for 5s. 
4. The `WalletLoading` one doesn't show any big decline, which is good. 
5. On average, the result fluctuations (err %) seemed to increase when the benchmarks ran for 5s compared to 10ms.

I believe that both the master & PR benchmarks shared in the PR description should be updated as they are more than 2 years old now, and a latest one would paint a more accurate picture. Also, the `WalletLoadingLegacy` benchmark can be removed from PR description as it no longer exists.

> Since imports can change the IsMine status of a transaction (although they can only be "promoted" from watchonly to spendable), all of the import RPCs will be a bit slower as they re-iterate all transactions and all outputs to update m_txos.

After the removal of the legacy wallets, there is just 1 import* RPC now - `importdescriptors`. Even though the overall results of the benchmarks are promising, I'm wondering if this PR should add `importdescriptors` benchmark for us to know what's the hit on this RPC. 

Slightly unrelated but now again that the legacy wallet is no more, the `*descriptor` benchmarks names can be updated to remove the descriptor suffix to ease it on the eyes - `WalletLoadingDescriptors` to `WalletLoading` - if I'm not missing anything.

<details>
<summary>

### Benchmarks (MacOS Sequoia)
</summary>

```
➜  bitcoin git:(master) ✗ ./build/bin/bench_bitcoin -filter="Wallet.*" -min-time=5000
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       60,048,154.50 |               16.65 |    0.9% |      1.33 | `WalletAvailableCoins`
|            8,084.01 |          123,701.05 |    1.1% |      5.51 | `WalletBalanceClean`
|           64,005.21 |           15,623.73 |    0.5% |      5.47 | `WalletBalanceDirty`
|            8,029.22 |          124,545.13 |    0.7% |      5.47 | `WalletBalanceMine`
|               33.16 |       30,153,409.91 |    0.8% |      5.31 | `WalletBalanceWatch`
|      745,284,217.00 |                1.34 |    1.6% |      8.17 | `WalletCreateEncrypted`
|      344,072,227.00 |                2.91 |    1.7% |      4.51 | `WalletCreatePlain`
|        1,265,623.20 |              790.12 |    1.5% |      0.07 | `WalletCreateTxUseOnlyPresetInputs`
|       97,241,712.60 |               10.28 |    1.7% |      5.37 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|               94.82 |       10,546,060.33 |    1.1% |      5.51 | `WalletIsMineDescriptors`
|              107.74 |        9,281,219.40 |    0.6% |      5.48 | `WalletIsMineMigratedDescriptors`
|      347,539,943.00 |                2.88 |    1.2% |      1.76 | `WalletLoadingDescriptors`
|    4,515,248,518.00 |                0.22 |    0.0% |      4.52 | `WalletMigration`

```
➜  bitcoin git:(wallet-txos) ✗ ./build/bin/bench_bitcoin -filter="Wallet.*" -min-time=5000
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       60,103,807.50 |               16.64 |    1.5% |      1.33 | `WalletAvailableCoins`
|            5,277.12 |          189,497.31 |    4.5% |      5.52 | `WalletBalanceClean`
|            5,669.83 |          176,372.17 |    1.4% |      5.69 | `WalletBalanceDirty`
|            5,161.86 |          193,728.76 |    1.6% |      5.57 | `WalletBalanceMine`
|               35.88 |       27,867,699.08 |    2.1% |      5.27 | `WalletBalanceWatch`
|      778,745,354.00 |                1.28 |    3.7% |      8.69 | `WalletCreateEncrypted`
|      350,162,310.00 |                2.86 |    1.7% |      3.95 | `WalletCreatePlain`
|        1,544,025.80 |              647.66 |    2.2% |      0.09 | `WalletCreateTxUseOnlyPresetInputs`
|      103,045,837.80 |                9.70 |    1.3% |      5.74 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|               96.31 |       10,382,776.81 |    2.0% |      5.48 | `WalletIsMineDescriptors`
|              116.97 |        8,549,245.41 |    0.9% |      5.51 | `WalletIsMineMigratedDescriptors`
|      352,261,692.00 |                2.84 |    0.6% |      1.76 | `WalletLoadingDescriptors`
|    4,671,199,177.00 |                0.21 |    0.0% |      4.67 | `WalletMigration`
</details>

rkrux reviewed (commented):
Benchmarks at 29cf059e6e592f7f82b982493a043219cdfa5b40

An update to my previous comment: https://github.com/bitcoin/bitcoin/pull/27286#pullrequestreview-2848375525

I dug a bit more and realised that in order to have a fair assessment between master and the PR, the `AvailableCoins` benchmark changes of the last commit 29cf059e6e592f7f82b982493a043219cdfa5b40 must be present in master as well so that both the results come from the same benchmark cases. I gcp'ed the commit onto master, removed the hardcoded `epochIterations` from `AvailableCoins` & `WalletCreateTx` functions like below, and let the benchmarks run for 5s each.

<details>
<summary>
wallet_create_tx.cpp diff
</summary>

```diff
--- a/src/bench/wallet_create_tx.cpp
+++ b/src/bench/wallet_create_tx.cpp
@@ -159,7 +159,7 @@ static void WalletCreateTx(benchmark::Bench& bench, const OutputType output_type
     if (coin_control.m_allow_other_inputs) target += 50 * COIN;
     std::vector<wallet::CRecipient> recipients = {{dest, target, true}};
 
-    bench.epochIterations(5).run([&] {
+    bench.run([&] {
         LOCK(wallet.cs_wallet);
         const auto& tx_res = CreateTransaction(wallet, recipients, /*change_pos=*/std::nullopt, coin_control);
         assert(tx_res);
@@ -198,7 +198,7 @@ static void AvailableCoins(benchmark::Bench& bench, const std::vector<OutputType
     auto bal = WITH_LOCK(wallet.cs_wallet, return wallet::AvailableCoins(wallet).GetTotalAmount()); // Cache
     assert(bal == 49 * COIN * (chain_size - COINBASE_MATURITY));
 
-    bench.epochIterations(2).run([&] {
+    bench.run([&] {
         LOCK(wallet.cs_wallet);
         const auto& res = wallet::AvailableCoins(wallet);
         assert(res.All().size() == (chain_size - COINBASE_MATURITY) * 2);
``` 
</details>

<details>
<summary>

#### Benchmark Results (best of the lot)
</summary>

```
➜  bitcoin git:(master) ✗ ./build/bin/bench_bitcoin -filter="Wallet.*" -min-time=5000
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       62,913,524.29 |               15.89 |    1.1% |      5.59 | `WalletAvailableCoins`
|            8,347.78 |          119,792.35 |    2.4% |      5.54 | `WalletBalanceClean`
|           64,450.99 |           15,515.66 |    0.5% |      5.53 | `WalletBalanceDirty`
|            8,474.18 |          118,005.46 |    3.0% |      5.18 | `WalletBalanceMine`
|               36.07 |       27,723,968.90 |    2.1% |      5.49 | `WalletBalanceWatch`
|      764,481,945.00 |                1.31 |    0.9% |      8.48 | `WalletCreateEncrypted`
|      360,601,555.00 |                2.77 |    0.9% |      4.49 | `WalletCreatePlain`
|        1,358,501.41 |              736.11 |    0.8% |      5.52 | `WalletCreateTxUseOnlyPresetInputs`
|      120,641,695.25 |                8.29 |    0.9% |      5.53 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|              114.60 |        8,725,874.44 |    0.8% |      5.58 | `WalletIsMineDescriptors`
|              113.78 |        8,789,224.41 |    0.2% |      5.30 | `WalletIsMineMigratedDescriptors`
|      366,820,995.00 |                2.73 |    0.5% |      1.83 | `WalletLoadingDescriptors`
|    5,000,027,548.00 |                0.20 |    0.0% |      5.00 | `WalletMigration`

```
➜  bitcoin git:(wallet-txos) ✗ ./build/bin/bench_bitcoin -filter="Wallet.*" -min-time=5000
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       60,758,414.00 |               16.46 |    0.8% |      5.55 | `WalletAvailableCoins`
|            5,369.55 |          186,235.52 |    0.5% |      5.51 | `WalletBalanceClean`
|            5,709.33 |          175,151.91 |    0.7% |      5.50 | `WalletBalanceDirty`
|            5,289.61 |          189,049.70 |    0.4% |      5.50 | `WalletBalanceMine`
|               38.12 |       26,233,493.24 |    2.5% |      5.33 | `WalletBalanceWatch`
|      753,308,252.00 |                1.33 |    0.5% |      8.28 | `WalletCreateEncrypted`
|      352,858,982.00 |                2.83 |    0.7% |      4.26 | `WalletCreatePlain`
|        1,392,278.42 |              718.25 |    2.2% |      5.21 | `WalletCreateTxUseOnlyPresetInputs`
|      103,969,867.00 |                9.62 |    0.7% |      5.53 | `WalletCreateTxUsePresetInputsAndCoinSelection`
|               97.30 |       10,278,004.67 |    0.7% |      5.56 | `WalletIsMineDescriptors`
|               96.68 |       10,343,863.28 |    0.6% |      5.49 | `WalletIsMineMigratedDescriptors`
|      361,421,296.00 |                2.77 |    1.4% |      1.81 | `WalletLoadingDescriptors`
|    4,682,511,111.00 |                0.21 |    0.0% |      4.68 | `WalletMigration`
</details>

### Observations
1. The `WalletBalance*` benchmarks show remarkable improvements like in the previous comment - except `WalletBalanceWatch` that sees a decline.
2. Unlike the previous case, both `WalletAvailableCoins` & `WalletCreateTxUsePresetInputsAndCoinSelection` have an improved perf now.
3. The `WalletIsMine*` benchmarks have also improved, though I have noticed fluctuations in results across reruns.
4. The `WalletLoading` one doesn't show a decline anymore, it seems stable across master & PR. Improved in some test runs like the one shared above.

<details open>
<summary>

##### Benchmarks percentage changes calculated based on op/s
</summary>

Benchmark | % change
-- | --
`WalletAvailableCoins` | 3.59
`WalletBalanceClean` | 55.47
`WalletBalanceDirty` | 1,028.87
`WalletBalanceMine` | 60.20
`WalletBalanceWatch` | -5.38
`WalletCreateEncrypted` | 1.53
`WalletCreatePlain` | 2.17
`WalletCreateTxUseOnlyPresetInputs` | -2.43
`WalletCreateTxUsePresetInputsAndCoinSelection` | 16.04
`WalletIsMineDescriptors` | 17.79
`WalletIsMineMigratedDescriptors` | 17.69
`WalletLoadingDescriptors` | 1.47
`WalletMigration` | 5.00
</details>

murchandamus reviewed (commented):
ReACK d73e4f878d143fe86d84e0e6f18c2a88c0cb8609

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/27286/commits/d73e4f878d143fe86d84e0e6f18c2a88c0cb8609

achow101 commented on src/wallet/transaction.h:
```diff
@@ -362,6 +363,30 @@ struct WalletTxOrderComparator {
+{
+private:
+    const CWalletTx& m_wtx;
+    const CTxOut& m_output;
```
The whole point is to not have `CWalletTx`s permanently in memory, so just an index pretty much defeats that.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -257,6 +257,10 @@ bool CachedTxIsFromMe(const CWallet& wallet, const CWalletTx& wtx, const isminef
     AssertLockHeld(wallet.cs_wallet);
+
+    // This wtx is already trusted
+    if (trusted_parents.contains(wtx.GetHash())) return true;
```
`trusted_parents` is not a global cache, it is a per-call cache, and is generally unique per-call to `CachedTxIsTrusted`. Since each call to `CachedTxIsTrusted` is guarded by `cs_wallet`, there can be no state changes made to any of the transactions. While that lock is held by the caller of `CachedTxIsTrusted`, the result of checking whether a specific transaction will always be the same, regardless of any state changes that may have occurred.

davidgumberg reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/27286/commits/39734259fc10b1339db2ae80c5fab12916cf0aff

Left a lot of non-blocking nits/questions/observations, many of which are out-of-scope, and all of which (except for commit hygiene nits) can be addressed in a follow up.

Benchmarked this branch with the hard-coded `nEpochIterations` removed as suggested by @rkrux (https://github.com/bitcoin/bitcoin/pull/27286#pullrequestreview-2850393076) against master with this commit cherry-picked on top, with `nEpochIterations` removed, and got the following results:

```console
$ ./build/bin/bench_bitcoin -min-time=10000 '-filter=Wallet(Available|CreateTxUse).*'
``` 

### PR branch
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       47,042,915.76 |               21.26 |    0.1% |  792,728,393.18 |  162,978,291.67 |  4.864 |  13,610,250.59 |    0.1% |     11.00 | `WalletAvailableCoins`
|          910,290.93 |            1,098.55 |    0.1% |   16,006,199.89 |    3,166,046.15 |  5.056 |     259,763.75 |    0.5% |     11.02 | `WalletCreateTxUseOnlyPresetInputs`
|       33,100,784.93 |               30.21 |    0.4% |  375,640,418.09 |  114,901,658.67 |  3.269 |  67,268,574.55 |    0.4% |     10.78 | `WalletCreateTxUsePresetInputsAndCoinSelection`

### master
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       46,834,069.48 |               21.35 |    0.0% |  792,798,330.50 |  162,224,895.00 |  4.887 |  15,794,362.95 |    0.1% |     10.96 | `WalletAvailableCoins`
|          929,561.19 |            1,075.78 |    0.3% |   16,460,517.63 |    3,232,134.94 |  5.093 |     293,191.79 |    0.4% |     11.04 | `WalletCreateTxUseOnlyPresetInputs`
|       41,569,035.73 |               24.06 |    0.2% |  517,355,083.92 |  144,349,136.46 |  3.584 |  91,003,243.17 |    0.2% |     11.00 | `WalletCreateTxUsePresetInputsAndCoinSelection`

`WalletCreateTxUsePresetInputsAndCoinSelection` is substantially faster, but `AvailableCoins()` looks similar.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2348,6 +2353,9 @@ util::Result<void> CWallet::RemoveTxs(WalletBatch& batch, std::vector<Txid>& txs
                 mapTxSpends.erase(txin.prevout);
+            for (unsigned int i = 0; i < it->second.tx->vout.size(); ++i) {
+                m_txos.erase(COutPoint(Txid::FromUint256(hash), i));
+            }
```
> I wonder what you think of e.g. making a bespoke class that houses m_txos and mapWallet where RefreshTXOsFromTx, RemoveTx etc. would live, and one could always be sure that the wallet's tracked transactions and tracked outputs always stay in sync. I admit that this scatters the code still, just in another way, but I wonder if at this size, we have to start more aggressively encapsulating functionality.

I don't think that makes sense to do as it would end up being just another wrapper around a `std::map`.

In any case, definitely out of scope for this PR.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                if (wallet.IsTxImmatureCoinBase(txo.GetWalletTx()) && txo.GetWalletTx().isConfirmed()) {
+                    ret.m_mine_immature += credit_mine;
+                    ret.m_watchonly_immature += credit_watchonly;
+                } else if (is_trusted && tx_depth >= min_depth) {
```
It would make more sense to drop unconfirmed balances when min_depth > 0 rather than include less than min_depth as a separate balance. Regardless, this PR is not trying to change user facing behavior, so this is out of scope.

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
-                ret.m_mine_untrusted_pending += tx_credit_mine;
-                ret.m_watchonly_untrusted_pending += tx_credit_watchonly;
+        for (const auto& [outpoint, txo] : wallet.GetTXOs()) {
+            Assert(MoneyRange(txo.GetTxOut().nValue));
```
Done

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -330,31 +345,19 @@ std::map<CTxDestination, CAmount> GetAddressBalances(const CWallet& wallet)
+            CAmount n = wallet.IsSpent(outpoint) ? 0 : txo.GetTxOut().nValue;
+            balances[addr] += n;
         }
     }
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -330,137 +330,146 @@ CoinsResult AvailableCoins(const CWallet& wallet,
         }
 
-        if (nDepth < min_depth || nDepth > max_depth) {
+        if (wallet.IsTxImmatureCoinBase(wtx) && !params.include_immature_coinbase)
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -330,137 +330,146 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+    // Cache for whether each tx passes the tx level checks (first bool), and whether the transaction is "safe" (second bool)
+    std::unordered_map<uint256, std::pair<bool, bool>, SaltedTxidHasher> tx_safe_cache;
+    for (const auto& [outpoint, txo] : wallet.GetTXOs()) {
+        const CWalletTx& wtx = txo.GetWalletTx();
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -330,137 +330,146 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-        for (unsigned int i = 0; i < wtx.tx->vout.size(); i++) {
-            const CTxOut& output = wtx.tx->vout[i];
-            const COutPoint outpoint(txid, i);
+        assert(mine != ISMINE_NO);
```
Done

davidgumberg commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                    ret.m_watchonly_immature += credit_watchonly;
+                } else if (is_trusted && tx_depth >= min_depth) {
+                    ret.m_mine_trusted += credit_mine;
+                    ret.m_watchonly_trusted += credit_watchonly;
```
My mistake, that makes sense.

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/ec25de35b2a51b493d36452f75096f790230a4e9

Reviewed range-diff (`git range-diff 3973425...ec25de35`), addresses all the nits I left in my previous review that were correct.  Mainly, the benchmark iterations are no longer unnecessarily hardcoded, two transaction level checks that were being performed per-output in `AvailableCoins()` are now part of the cached tx-level checks, and stylistically, code was moved to more closely resemble the order of checks in master, making the (whitespace-hidden) diff much smaller, and the changes in this PR easier to remove in my opinion: e.g. (https://github.com/bitcoin/bitcoin/commit/854241298059f8b56563ca7ca64f5b61b59f119b?w=1 vs https://github.com/bitcoin/bitcoin/pull/27286/commits/2703d84b7efafcce05387a5e018363c3c59c4b58?w=1)

I benchmarked again (w/ different hardware[^1])  and it looks like the performance gap in the  `AvailableCoins()` benchmark is improved over the earlier branch (10% faster than master vs 5% faster than master) 

<details>

<summary>

### Benchmark results

</summary>

```console
$ ./build/bin/bench_bitcoin -min-time=30000 '-filter=Wallet(Available|CreateTxUse).*'
```

#### pr branch (https://github.com/bitcoin/bitcoin/commit/ec25de35b2a51b493d36452f75096f790230a4e9)

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       39,666,039.18 |               25.21 |    0.2% |  743,701,163.19 |  184,433,998.97 |  4.032 |  13,191,744.00 |    0.1% |     32.98 | `WalletAvailableCoins`
|          818,711.43 |            1,221.43 |    0.1% |   16,376,793.42 |    3,821,152.74 |  4.286 |     264,358.23 |    0.2% |     33.07 | `WalletCreateTxUseOnlyPresetInputs`
|       28,955,730.17 |               34.54 |    0.7% |  370,601,527.15 |  135,128,956.19 |  2.743 |  66,153,487.66 |    0.4% |     33.26 | `WalletCreateTxUsePresetInputsAndCoinSelection`

#### master (e2174378aa8a3 + a39074c52afd073abd9101dc6) 

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       44,211,982.81 |               22.62 |    0.2% |  803,845,700.31 |  205,309,518.10 |  3.915 |  18,081,465.08 |    0.0% |     33.04 | `WalletAvailableCoins`
|          855,419.56 |            1,169.02 |    0.1% |   16,581,860.02 |    3,991,552.91 |  4.154 |     355,371.44 |    0.2% |     33.02 | `WalletCreateTxUseOnlyPresetInputs`
|       35,074,882.27 |               28.51 |    0.2% |  505,591,550.24 |  163,578,562.91 |  3.091 |  91,987,283.54 |    0.2% |     33.16 | `WalletCreateTxUsePresetInputsAndCoinSelection`

#### old pr branch (https://github.com/bitcoin/bitcoin/commit/39734259fc10b1339db2ae80c5fab12916cf0aff) with `epochIterations` removed

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       42,297,947.25 |               23.64 |    0.0% |  774,588,933.86 |  196,770,180.57 |  3.937 |  15,813,962.88 |    0.1% |     32.95 | `WalletAvailableCoins`
|          852,312.97 |            1,173.28 |    0.2% |   16,650,566.31 |    3,978,823.89 |  4.185 |     374,875.56 |    0.2% |     33.04 | `WalletCreateTxUseOnlyPresetInputs`
|       28,841,852.53 |               34.67 |    0.4% |  373,458,753.83 |  134,591,208.06 |  2.775 |  68,667,702.77 |    0.4% |     33.26 | `WalletCreateTxUsePresetInputsAndCoinSelection`

</details>

[^1]: AMD Ryzen 9 7900X 12-Core Processor used in these benchmarks.

rkrux reviewed (commented):
Perf Review ec25de3

Unfortunately, I have not been able to do a code review of this PR yet but I wanted to ensure that I finish the performance review of this change; I will able to do a code review in a week or two.

While benchmarks are reliable enough to give an idea about the performance, I wanted to contextualise the findings of those benchmarks by calling the related RPCs, and also wanted to address the concern I had about additional memory usage.

Motivated by [this issue comment](https://github.com/bitcoin/bitcoin/issues/27002#issuecomment-1412386397) from achow, I created a perf test in my regtest using [this script](https://github.com/rkrux/bitcoincore_prnotes/blob/main/27286-wallet-txos/script.py) by generating a wallet with 7,117 UTXOs among 35,744 Bech32 addresses spanning across 362,621 wallet transactions generated by 7370 blocks. After generating the initial `coinbase` transactions, I created `passthrough` transactions that let the unspents flow through them - every such transaction spent 1 unspent and sent 99.5% of the amount to a new wallet address and the remaining amount equally divided among 10 non-wallet addresses. I did this to simulate the scenario where the user is sending funds to different addresses and the rest is sent back to the wallet as change.  

**Results:**
<img width="985" alt="perf-test-result" src="https://github.com/user-attachments/assets/ea8a6ce5-7d68-4faf-8633-55ee60aba1fc" />
_System - MacOS Sequoia, 2.6 GHz 6-Core Intel Core i7, 16GB Memory_

**Observations:**
1. As conveyed by the benchmarks as well, there is remarkable improvement in the `getbalances` RPC latency.
2. `listunspent` RPC also has a reduced latency. The `send` RPC, which relates to the `WalletCreateTxUsePresetInputsAndCoinSelection` benchmark also improved considerably in the second run, though it took slightly more time in the first run.
3. I expect the last 3 RPCs listed to further improve after #27865 when unspendable TXOs are separated from the spendable ones.
4. I didn't expect a lot of memory consumption increase because the `WalletTXO` object contains only references to `CWalletTx` & `CTxOut`, and, moreover, the `m_ismine` member will also be removed after #32523. This expectation was met in this perf test as there is not a lot of increase in the `maximum resident set size` parameter of the `time` command.
5. The consistent tracking of the `top -pid $(pgrep bitcoind)` command _after_ the RPCs execution showed me that roughly a ~20MB increment was there in this perf test compared to master in this wallet having more than 7k unspents in ~360k transactions.

**Conclusion:**
While the results show data from 2 runs, more runs might form consistency in the metrics delta of `loadwallet` & `listtransactions` RPCs. The `listunspent` & `getbalances` results already show consistent improvement to me; the memory consumption increase is also not significant, which is good imo. I suppose the changes of this PR furthered by #27865 and #32523 should bring in more improvement. 

So, Concept ACK ec25de3.

------

**Edit:** Apparently, my test has ~37k transactions and not ~360k. I used the output of `listtransactions` RPC but realised the difference later when I checked the `txcount` output of `getwalletinfo` RPC.

murchandamus commented on src/wallet/receive.cpp:
```diff
@@ -297,27 +298,41 @@ bool CachedTxIsTrusted(const CWallet& wallet, const CWalletTx& wtx)
+                }
+
+                // Set the amounts in the return object
+                if (wallet.IsTxImmatureCoinBase(wtx) && wtx.isConfirmed()) {
+                    ret.m_mine_immature += credit_mine;
+                    ret.m_watchonly_immature += credit_watchonly;
+                } else if (is_trusted && tx_depth >= min_depth) {
+                    ret.m_mine_trusted += credit_mine;
+                    ret.m_watchonly_trusted += credit_watchonly;
+                } else if (!is_trusted && wtx.InMempool()) {
+                    ret.m_mine_untrusted_pending += credit_mine;
+                    ret.m_watchonly_untrusted_pending += credit_watchonly;
```
I was just rereading the commit "wallet: Change balance calculation to use m_txos" (dde7cbe105ba6daaa636466d0fa3a83d15609417) and I’m aware that this is not a behavior change, but I am wondering about how the balances are calculated.

As seen, this distinguishes three groups of TXOs:

- "immature"
  `if (wallet.IsTxImmatureCoinBase(wtx) && wtx.isConfirmed()) {`
- "trusted"
  `} else if (is_trusted && tx_depth >= min_depth) {`
- "untrusted_pending"
  `} else if (!is_trusted && wtx.InMempool()) {`
  
and it made me wonder, whether a change output on an unconfirmed transaction our wallet created would just not count toward any of these balances. We trust our own transactions, so our change output should be `is_trusted`—and `min_depth` is 0 by default. Nevermind.

murchandamus reviewed (commented):
reACK 215e5999e20

ishaanam reviewed (commented):
reACK 215e5999e2070a38c68e343c5c3f1dc37d567f58

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/27286/commits/215e5999e2070a38c68e343c5c3f1dc37d567f58
